packages/glimmer-runtime/lib/compiled/opcodes/content.ts

```diff
@@ -127,3 +120,50 @@ export class UpdateTrustingAppendOpcode extends UpdatingContentOpcode {
     return { guid, type, details };
   }
 }
+
+export class UpdateCautiousAppendOpcode extends UpdatingContentOpcode {
+  type = 'update-cautious-append';
+  private cache: ReferenceCache<string>;
+  private bounds: Fragment;
+  private textNode: Text;
+
+  constructor(cache: ReferenceCache<string>, bounds: Fragment, textNode: Text) {
+    super();
+    this.cache = cache;
+    this.bounds = bounds;
+    this.textNode = textNode;
+  }
+
+  evaluate(vm: UpdatingVM) {
+    let value = this.cache.revalidate();
+
+    if (isModified(value)) {
+      let parent = <HTMLElement>this.bounds.parentElement();
+
+      if (isSafeString(value)) {
+        this.textNode = null;
+        let nextSibling = clear(this.bounds);
+        this.bounds.update(vm.dom.insertHTMLBefore(parent, nextSibling, value));
+      } else {
+        if(this.textNode) {
+          this.textNode.nodeValue = value;
+        } else {
+          let nextSibling = clear(this.bounds);
+          this.bounds.update(vm.dom.insertTextBefore(parent, nextSibling, value));
+          this.textNode = this.bounds.firstNode() as Text;
```

Comment:
I avoided doing this because nowhere else are bounds constructed inside of `content.ts` -- they seem to all be constructed at a different layer (inside of the `dom` object). I can move it, it just seemed inconsistent.